src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -389,25 +376,26 @@ private void createUserAppsEntitiesAndAssociations(
                 Set<DataEdgeKey> dataEdgeKeys = new HashSet<>();
 
                 String appPackageName, appName;
-                appPackageName = appName = appEntity.get( propertyTypeIdsByFQN.get( FULL_NAME_FQN ) ).iterator().next()
-                        .toString();
+                appPackageName = appName = getFirstValueOrNullByUUID( appEntity, FULL_NAME_FQN );
                 if ( systemAppPackageNames.contains( appPackageName ) )
                     continue; // a 'system' app
 
                 if ( appEntity.containsKey( propertyTypeIdsByFQN.get( TITLE_FQN ) ) ) {
-                    appName = appEntity.get( propertyTypeIdsByFQN.get( TITLE_FQN ) ).iterator().next().toString();
+                    appName = getFirstValueOrNullByUUID( appEntity, TITLE_FQN );
                 }
 
-                String dateLogged = getMidnightDateTime( appEntity.get( propertyTypeIdsByFQN.get( DATE_LOGGED_FQN ) )
-                        .iterator().next()
-                        .toString() );
+                String dateLogged = getMidnightDateTime( getFirstValueOrNullByUUID( appEntity, DATE_LOGGED_FQN ) );
 
                 // create entity in chronicle_user_apps
                 Map<UUID, Set<Object>> userAppEntityData = new HashMap<>();
-                userAppEntityData.put( propertyTypeIdsByFQN.get( FULL_NAME_FQN ), Sets.newHashSet( appPackageName ) );
-                userAppEntityData.put( propertyTypeIdsByFQN.get( TITLE_FQN ), Sets.newHashSet( appName ) );
+                userAppEntityData.put( propertyTypeIdsByFQN.get( FULL_NAME_FQN ), ImmutableSet.of( appPackageName ) );
+                userAppEntityData.put( propertyTypeIdsByFQN.get( TITLE_FQN ), ImmutableSet.of( appName ) );
+
+                EntityKey userAppEK = getUserAppsEntityKey( userAppEntityData );
+                UUID userAppEntityKeyId = entityKeyIdMap.getOrDefault( userAppEK, null );
+                if ( userAppEntityKeyId == null )
+                    continue;
```

Comment:
Yes, seems like the right thing to do is to skip the current entity and process the next. But `userAppEntityKeyId` is guaranteed to not be null since the EntityKey generated here is exactly the same as the one generated when the `entityKeyId` map is created. Therefore this `continue` is just a fail-safe.